File: 算法/回溯/剑指offer12矩阵中的路径.py
# DFS
# 上下左右 四格
from typing import List

coor = [[-1, 0], [1, 0], [0, -1], [0, 1]]


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:

        if not board or not word:
            return False

        start_coors = []
        num_row = len(board)
        num_col = len(board[0])

        vis = [[False] * num_col for i in range(num_row)]

        for i in range(num_row):
            for j in range(num_col):
                if board[i][j] == word[0]:
                    start_coors.append((i, j))

        for i, j in start_coors:
            vis[i][j] = True
            result = self.dfs(board, word, vis, i, j)
            vis[i][j] = False
            if result:
                return True

        return False

    def dfs(self, board, word, vis, i, j):
        if not word:
            return False
        num_row = len(board)
        num_col = len(board[0])

        if board[i][j] == word[0]:
            if len(word) == 1:
                return True
            next_word = word[1:]

            for c in coor:
                next_i = i + c[0]
                next_j = j + c[1]
                if next_i >= 0 and next_i < num_row and next_j >= 0 and next_j < num_col and (
                        not vis[next_i][next_j]
                ) and board[next_i][next_j] == next_word[0]:

                    vis[next_i][next_j] = True
                    res = self.dfs(board, next_word, vis, next_i, next_j)
                    vis[next_i][next_j] = False
                    if res:
                        return res
            return False
        else:
            return False

File: 算法/回溯/剑指offer12矩阵中的路径.cpp
#include <bits/stdc++.h>
using namespace std;

int coors[4][2] = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};

class Solution
{
public:
    bool exist(vector<vector<char>> &board, string word)
    {
        if (board.empty() || board[0].empty() || word.empty())
        {
            return false;
        }
        int num_row = board.size();
        int num_col = board[0].size();

        for (int i = 0; i < num_row; ++i)
        {
            for (int j = 0; j < num_col; ++j)
            {
                if (board[i][j] == word[0])
                {
                    if (dfs(board, word, i, j, 0))
                    {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    bool dfs(vector<vector<char>> &board, string &word, int i, int j, int word_idx)
    {
        if (word_idx == (word.size() - 1))
        {
            return true;
        }

        int next_word_idx = word_idx + 1;
        char origin_char = board[i][j];

        int next_i, next_j;

        board[i][j] = '/';
        for (int idx = 0; idx < 4; ++idx)
        {
            next_i = i + coors[idx][0];
            next_j = j + coors[idx][1];

            if (next_i >= 0 && next_i < board.size() && next_j >= 0 && next_j < board[0].size() && board[next_i][next_j] == word[next_word_idx])
            {
                if (dfs(board, word, next_i, next_j, next_word_idx))
                {
                    return true;
                }
            }
        }
        board[i][j] = origin_char;
        return false;
    }
};